### PR TITLE
[kubeadm] warn if the CIDR is missing

### LINKED ISSUE #1011: kubeadm should warn about missing podSubnet/clusterCIDR
# BUG REPORT

## Versions

**kubeadm version** (use `kubeadm version`):
kubeadm version: &version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:14:41Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
**Environment**:
- **Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:17:28Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.0", GitCommit:"91e7b4fd31fcd3d5f436da26c980becec37ceefe", GitTreeState:"clean", BuildDate:"2018-06-27T20:08:34Z", GoVersion:"go1.10.2", Compiler:"gc", Platform:"linux/amd64"}
- **Cloud provider or hardware configuration**:
brightbox
- **OS** (e.g. from /etc/os-release):
Ubuntu 18.04 LTS
- **Kernel** (e.g. `uname -a`):
Linux srv-jlhyq 4.15.0-23-generic #25-Ubuntu SMP Wed May 23 18:02:16 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
- **Others**:

## What happened?

Without a pod subnet specified, or a specific clusterCIDR specified you get: 

```
W0720 13:06:36.713799       1 proxier.go:311] clusterCIDR not specified, unable to distinguish between internal and external traffic
```

from kube-proxy and masquerade facilities will be affected

## What you expected to happen?

The clusterCIDR should be given to kube-proxy, or a warning issued during init

Given the number of CNI plugins that require a subnet specification there should be at least a warning from kubeadm about the impact of failing to tell the system what addresses the pods will be using. 

## How to reproduce it (as minimally and precisely as possible)?

kubeadm init without a podSubnet

## Anything else we need to know?

There appears to be a bit of a battle going on between the top down allocation of system's addresses and the bottom up discovery of the system's addresses, with bits of each leaking into the instructions. Perhaps a pre-flight check is required to ensure consistency.

### PR DESCRIPTION
**Pull request by xlgao-zju**
**What this PR does / why we need it**:
warn if the CIDR is missing

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes [kubernetes/kubeadm/#1011](https://github.com/kubernetes/kubeadm/issues/1011)

**Special notes for your reviewer**:
/assign @stealthybox 
/assign @neolit123 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by neolit123** [comment]
@xlgao-zju thanks!
seems correct as per the issue.

LGTM

/kind bug
/test pull-kubernetes-e2e-kops-aws
/assign @timothysc

**Review by fabriziopandini — APPROVED** [review]
@xlgao-zju thanks for the fix!
/approve 
I let final lgtm to @neolit123 or @stealthybox who followed the issue

**Comment by xlgao-zju** [comment]
Ping @neolit123 for `lgtm` label, since you have already given your LGTM.

**Comment by neolit123** [comment]
/hold
something else to note here.

i tested locally.
this will always show a preflight warning because we don't have a default `podSubnet` value in the config (which is understandable), but we do always set `serviceSubnet` to `10.96.0.0/12`.

so the question is - is it OK to show this warning?
or should we rather default to some value?

the verification is performed here:
https://github.com/kubernetes/kubernetes/blob/2f63238c478687276f9ce6d56f1364a0012275a6/cmd/kubeadm/app/preflight/checks.go#L869

/assign @kad 
for opinion on the subject.

**Comment by stealthybox** [comment]
This is a good first stab at it @xlgao-zju.
The output of your patch currently looks like this.
```
kubeadm init
[init] using Kubernetes version: v1.11.3
[preflight] running pre-flight checks
	[WARNING HTTPProxyCIDR]: the CIDR is missing
[preflight/images] Pulling images required for setting up a Kubernetes cluster
[preflight/images] This might take a minute or two, depending on the speed of your internet connection
```

I recommend separating the CIDRMissing checks into their own dedicated checks instead of reusing the nil case for the Proxy check.

There should be a dedicated check for both `cfg.Networking.ServiceSubnet` and `cfg.Networking.PodSubnet`, so that the user can determine which fields they're being warned about and why they might want to define them explicitly.

**Comment by stealthybox** [comment]
@neolit123 @kad I agree that it would be preferable to have default values for both subnets.

Poking around, this doesn't seem required to make kube-proxy work, but if we can default the `Notworking.PodSubnet` and set the kube-proxy `--cluster-cidr` all the time, we should.

**Comment by neolit123** [comment]
what value should we default to PodSubnet?

**Comment by dims** [comment]
@timothysc @neolit123 what's the word on this? (needed for 1.12?)

**Comment by neolit123** [comment]
the PR is needed for 1.12 but it also needs discussion.
realistically, it might go in 1.12.1.

**Comment by kad** [comment]
I don't think we really need that check. Or, we can do it differently. If that parameter is needed, let's create a separate check for presence of CIDR parameter instead of trying to hide it inside HTTP Proxy CIDR check.

**Comment by xlgao-zju** [comment]
I think we can set the default value. And how about doing this check before using the default value. So users will know that we will use the default value, because he/she did not specify the CIDR.

**Comment by neolit123** [comment]
we should set a default value in all cases as we do for the service CIDR.

i think there are two options here:
A. set a default value, but only warn if the user removes it and makes the value empty.
(this is what we do for service CIDR)
B. (@xlgao-zju this is your suggestion) warn if the user does not set the pod CIDR and tell that a default value will be used.

i'm leaning towards A, but feedback  from others will be appreciated.
but we will probably delay this after 1.12.

**Comment by pmichali** [comment]
For bridge and PTP plugins, a podSubnet is not used, right? Are we considering that case?

**Comment by stealthybox** [comment]
Considering @pmichali's point about clearly supporting the bridge plugin, I propose:

When the podSubnet is empty, we should:
- notify the user it's empty
- indicate the general expected use-case(s)
- indicate that kube-proxy and any other components are not aware of podIP's -- (need further clarification here on all sensible component configurations)

**If** we default the podSubnet to some value, we should:
- notify the user of the default
- indicate the general expected use-case(s)
- indicate which use-cases don't function or benefit

___________________
The current default behavior is to leave the podSubnet implicitly empty.
We need to decide if we want to provide a default value for it.

A fair argument for providing a default is VXLAN CNI setups are common and will benefit from the option.
However, it's a breaking change for existing configs.

**Comment by kad** [comment]
IMHO: keep value as currently empty. Create new check that will warn about cases where it will be problematic.

**Review by kad — CHANGES_REQUESTED** [review]
Create separate check for empty CIDRs if needed.

**Comment by bart0sh** [comment]
@kubernetes/sig-cluster-lifecycle-pr-reviews this PR seems outdated? Should we close it?

**Comment by neolit123** [comment]
the original ticket is frozen and we didn't decide on the correct path for this.
/close

**Pull request closed without merging**
